.github/workflows/upgrade.yaml

```diff
@@ -0,0 +1,53 @@
+name: Upgrade Juju Controller and Juju Model
+
+on:
+  workflow_dispatch:
+    inputs:
+      agent_version:
+        description: 'Version to upgrade to'
+        required: true
+
+jobs:
+  build:
+    name: Trigger Juju ugrade controller and model on EKS Deployment
+    runs-on: ubuntu-latest
+    steps:
+      - name: Installing Dependencies
+        run: |
+          sudo snap install juju --channel=3.1/stable --classic
+      - name: Install Docker
+        uses: docker-practice/actions-setup-docker@master
+
+      # We need the kubeconfig and the controllers.yaml file into our environment
+      # in order to login and refresh the charms.
+      - name: Adding Credentials
+        shell: bash
+        run: |
+          mkdir -p ~/.kube
+          mkdir -p ~/.local/share/juju
+          echo "${KUBECONFIG_B64}" | base64 -d > ~/.kube/config
+          echo "${CONTROLLERS_B64}" | base64 -d > ~/.local/share/juju/controllers.yaml
+          # echo "${CONTROLLER_PASSWORD}" | juju login -c finos-legend-v3 -u admin
+        env:
+          KUBECONFIG_B64: "${{ secrets.KUBECONFIG_B64 }}"
+          CONTROLLERS_B64: "${{ secrets.CONTROLLERS_B64 }}"
+          CONTROLLER_PASSWORD: "${{ secrets.CONTROLLER_PASSWORD }}"
+
+      - name: Upgrade Juju Controller and Juju Model
+        run: |
+          # Compare if the juju version is lower than the version we want to upgrade to
+          if dpkg --compare-versions $(juju version | cut -d- -f1) "lt" ${{ github.event.inputs.agent_version }}
```

Comment:
So, a few things here to consider. You're comparing the juju client version to the ``agent_version`` given. That is not going to ensure that the controller is always going to be upgraded to a newer version. The juju client version != juju controller version.

Then, as a suggestion, I think it would be simpler to treat the oposite case: if the controller version is newer than the ``agent_version``, ``echo  "Upgrade skipped...`` and exit 1 afterwards (technically, it should end on a failure, since the input was invalid, so the upgrade basically failed).

Reply:
Updated it to take the `controller-model-version` and check if that one is greater than the `agent_version`